Load variables from bash script generated by cloud-init

jsf9k opened the pull request:
## 🗣 Description

This pull request:
* Combines the master and replica FreeIPA setup scripts into a single script.
* Reads non-sensitive variable values from a small bash file that will be generated by cloud-init.

## 💭 Motivation and Context

* I merged the master and replica setup scripts into a single file because doing so gets rid of some code duplication.
* I read the non-sensitive variable values from a bash file generated by cloud-init to make installation as easy and hands-off as possible.  The user still has to enter sensitive data such as the Kerberos password for the `admin` user.

See also cisagov/freeipa-server-tf-module#26 and cisagov/cool-sharedservices-freeipa#14.

## 🧪 Testing

I have used this code to successfully redeploy the FreeIPA server cluster in our staging environment.

## 🚥 Types of Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (causes existing functionality to change)

## ✅ Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

jsf9k committed:
Move server and replica install scripts to /usr/local/sbin

This is a better place for sensitive scripts like this to live.

jsf9k committed:
Remove unnecessary ipa-certupdate commands

jsf9k committed:
Load variables from a bash script created by cloud-init

jsf9k committed:
Combine the replica and server scripts into a single file

This gets rid of some duplicated code.

jsf9k committed:
Add code to remove passwords from vars file generated by cloud-init

jsf9k committed:
Rename "server" installation type to "master"

"server" is pretty ambiguous.

jsf9k committed:
Assume the user will enter the necessary credentials

The PWs will no longer be stored in the bash variables script created
by cloud-init.

Also:
* Go ahead and setup KRA in case we ever want to use it
* Add the --mkhomedir option
* Remove the --unattended option, since the user is expected to enter
his or her own credentials
* Add a principal alias for the instance ID so that folks can ssh in
via SSM Session Manager, authenticating with their Kerberos
credentials

jsf9k committed:
Remove extra line mistakenly left in after cut and paste

jsf9k committed:
Fix typo

jsf9k committed:
Remove PTR record jazz and run ipa-client-install before ipa-replica-install

We are no longer trying to install the replica at boot, so there is no
need to wait until the PTR record is in place.  By the time the
terraform apply completes, the PTR record is guaranteed to be in
place.

Also remove the loop that waits until the master is up and running.
Since we would be setting up the servers manually, the master will be
completely set up before you even log into the replica.  This allows
us to stop passing master_hostname as a variable.

Finally, we run ipa-client-install before running
ipa-replica-install.  For some reason ipa-replica-install does not
appear to pass the value associated with the --principal flag to
ipa-client-install, so this is necessary.

jsf9k committed:
Run kinit admin before running ipa certmaprule-add

When installing the master (the first IPA server), we need Kerberos
credentials before running this command.  kinit will prompt the user
fro the admin password before proceeding.

jsf9k committed:
Move the IP address foo into the master case label

It is not used if we are installing a replica.

jsf9k committed:
Add missing "host/" to start of host principal alias

dav3r commented on files/setup_freeipa.sh:
```diff
@@ -0,0 +1,127 @@
+                               --mkhomedir
+
+            # Get kerberos credentials and create the dhs_certmapdata
+            # rule
```
Can this comment be expanded slightly to explain why we add the `dhs_certmapdata` rule?

jsf9k replied:
See commit 660bf23.

dav3r replied:
Thanks for adding that excellent explanation!

dav3r commented on files/setup_freeipa.sh:
```diff
@@ -0,0 +1,127 @@
+}
+
+
+if [ $# -lt 1 ]
```
Wouldn't `if [ $# -ne 1 ]` be better here?

jsf9k replied:
See commit 372efa8.

dav3r reviewed (approved):
Very good old chap! 🎩 
I have but two brief comments for your consideration.

jsf9k committed:
Explain why the ipa certmaprule-add commend is necessary

jsf9k committed:
Specify that the script have precisely one argument

The previous code required _at least_ one argument and then ignored
the extras.

felddy commented on files/setup_freeipa.sh:
```diff
@@ -0,0 +1,141 @@
+}
+
+# Get the IP address corresponding to an interface
+function get_ip {
+    ip --family inet address show dev "$1" | \
+        grep inet | \
+        sed "s/^ *//" | \
+        cut --delimiter=' ' --fields=2 | \
+        cut --delimiter='/' --fields=1
```
I feel like this might come back to visit us later.

felddy replied:
`ip --family inet address show dev "$1" | grep -Po 'inet \K[\d.]+'`

Explained: https://regexr.com/56khv

Stolen from: https://unix.stackexchange.com/questions/87468/is-there-an-easy-way-to-programmatically-extract-ip-address

jsf9k replied:
See commit cefc6f7.

jsf9k committed:
Make curl be silent

The progress output is not really helpful or wanted in this context.

felddy reviewed (approved):
Looks good!  See idea about IP extraction from `ip`.  🌶️

jsf9k committed:
Simplify get_ip function

Thanks to @felddy for the suggestion.  This solution is probably also
more future-proof than what I had before.
